### DISCUSSION TITLE
Higher precision of expense rates

### DISCUSSION THREAD

**Discussion by ljbs** [kimai/kimai#5161] [General]
**Use case**
We recently bought and installed the Expenses plugin on our self-hosted Kimai instance and we're very happy about it so far. However, we face an issue that is (strongly) related to #3592 and #5114: we would need up to 4 decimals rather than having them rounded up to 2 decimals. Our specific use case is about tracking mileage reimbursements (and exporting them for accounting purposes), the rates of which are locally regulated by the government as fixed rates per distance unit (km or mi) with -you guessed it- 4 decimals. 

**Current workaround**
As a current workaround (hack, rather), I manually replaced the rounded rate of the relevant expense category in the SQL database to its exact value. That works fine upon submitting a new expense, but obviously no longer once the data of the expense record are edited and all numbers are re-calculated by Kimai. 

**Question**
I do realize we can externally implement some postprocessing steps to correct for the rounding after exporting all data, but since having everything as a single source of truth in the Kimai environment is very neat, it would be cool being able to deal with this in a somewhat 'clean' way. Any suggestion, workaround or pointer that could bring us in that direction would be greatly appreciated! 

**Similar discussions and issues**
* #5114 
* #3539
* #3592

**Comment by kevinpapst** [comment]
And the next idiotic government decides to use 5 decimals... there is nothing like € 0.5555, its either 0.55 or 0.56.
Do you think that any tax auditor in this world would cause troubles if you use either of those?


Edit: btw - thanks for using Kimai! 😄

**Reply by ljbs** [reply]
I'm inclined to say they'd even find something to nitpick about if it were exact 😉. Jokes aside, I get your point and fully understand it is impossible for the mainline framework to meet all these different (stupid) requirements - I was just wondering whether there would be a simple trick for us to make it work, and from your reply I conclude there isn't. Fair enough.

Thanks for your swift reply btw, much appreciated!

**Reply by kevinpapst** [reply]
No I mean honestly. Auditors are looking for invalid bookings, faked receipts, employees without registration ...  but not for some obscure rounding issue that causes a few € difference over the course of thousands of kilometers.
Especially if you decide to go with the lower version and use (in my made up example) 0.55 and not the 0.56 variant. 

I have the slight feeling you are trying to fix a non-existing issue here. Did you ever hear about any case of that?

Obviously I am not a tax consultant and my opinion is just that: thoughts of a developer and sole business owner. 

From the top of my head I don't know where rounding happens and how it is stored. 
I assume the database cannot hold 4 decimals, but you can post a Feature request at https://github.com/kimai/Paid-plugins and we see if someone else has the same issue.

**Reply by ljbs** [reply]
To be fair I don't know either, I'm a tech guy/co-owner trying to improve the administrative workflow by getting rid of the mess we are currently having with spreadsheets and deprecated forms based on templates our accountants provide. Basically I'm trying to integrate all those in Kimai now, which, to be clear, has already proven to be a major improvement in terms of user-friendliness 😅 . 

But you're right in the sense that all of this appears to be quite some fuzz for relatively small rounding errors, and I also believe that rounding down should not cause any discussions. I will talk to our accountants and see how they look at this - if deemed necessary, I'll make a feature request in the appropriate repo and reference it here.
